Pixel position mapping

owtscharenko opened the pull request:
I provided a method in `PixelModules::MakePositionMap` creating an unordered_map to hold the values. This needs to be called once before reconstruction, and is referred to in the `ShipPixelHit:EndPoints` method. Does it make sense like this and where would `MakePositionMap` be called from?

olantwin commented:
Hi Niko,
Apologies for the late reply.

Could you rebase the changes on master? There are many changes in the pull request which are unrelated, making it very hard to determine what you changed. 

I would do the following:

`ShipPixelHit::GetPixelXYZ()` (or something like that) is a method, which looks up the position in the map, and if it does not exists, creates it using `MakePositionMap`.
That way, the user doesn't have to know about how this is done behind the scenes.

Cheers,
Oliver

owtscharenko commented:
I rebased on your master branch.
Regarding you suggestion how to call the mapping function: I believe this is only possible in a superior script or class taking care of the reconstruction?
As far as I understand for each hit an instance of the ShipPixelHit class will be created, so how should these classes know about any object outside?
So I would put the code in the PixelModules class, but as far as I understand this not where code is actually supposed to be running but only a collection of methods to be called. Is this correct?

olantwin commented:
You can use [static members](https://en.cppreference.com/w/cpp/language/static) to share a member between all instances of the class.

I would probably make this a pointer, and make doubly sure that it's not serialised by ROOT.

olantwin commented:
Could you please rebase again? It seems the problem was that you included some of Thomas's and Antonio's commits, which I did not yet have in my master branch.
I just updated it, so that they are now included. You may need to remove them during interactive rebase.

owtscharenko commented:
I rebased again, and also implemented a solution in ShipPixelHit using static members, I do not know though whether the map will be actually initialized like this.

olantwin commented:
You'll probably want to create the map on first use, not before.

The way you implement if currently, when you e.g. convert the raw data, it will also create the map, as the class is included.

Also, you should add `//!` behind this pointer member when declaring it, so that ROOT know, that it should not be written to file.

olantwin commented:
See here: https://root.cern.ch/root/html/guides/users-guide/AddingaClass.html#adding-a-class-with-a-shared-library for the `//!` explanation.

olantwin commented on charmdet/ShipPixelHit.h:
```diff
@@ -16,7 +16,8 @@ struct HitID {
 
 class ShipPixelHit : public ShipHit {
-  static std::shared_ptr <std::unordered_map<int, TVector3>> PixelPositionMap;
+  static std::shared_ptr <std::unordered_map<int, TVector3>> PixelPositionMap; //
```
`//!`

Why a shared pointer?

Might be worth making the pointer and ony expose a read-only interface to python, to make sure nobody can mess up the map.

owtscharenko replied:
do you mean making the whole thing private?

olantwin replied:
No, just those two data members.
They might already be by default actually, but I would make it explicit by moving them to below the `private:` label.

owtscharenko commented:
done

olantwin commented on charmdet/ShipPixelHit.cxx:
```diff
@@ -1,6 +1,201 @@
+
+// compute PixelPositionMap once
+
+int ShipPixelHit::MapInitialized = 0;
```
This I would probably move to the header. Not entirely sure, whether it may be in the declaration itself.

We could also remove `MapInitialized` entirely and just use `nullptr` to determine whether the map is usuable.
What do you think?

Also, for a static pointer, I don't think `shared_ptr` adds any advantages, or am I missing anything? Or does it actually destruct the map if no hits are around anymore?

owtscharenko replied:
I will follow your suggestion, for the shared pointer, it was exactly my fear that the map might be destructed. But I do not truly know what will happen.
If you want I can change it to a static pointer and we will see what happens.

owtscharenko replied:
It seems that a pointer is initialized to `null`, so I now check for the boolean value of the pointer itself. This gets rid of `MapInitialized`

olantwin replied:
> I will follow your suggestion, for the shared pointer, it was exactly my fear that the map might be destructed. But I do not truly know what will happen.
> If you want I can change it to a static pointer and we will see what happens.

As the shared pointer never goes out of scope, I think it would never be destructed, which should be the same as for the normal ptr.

olantwin commented on charmdet/ShipPixelHit.cxx:
```diff
@@ -1,6 +1,201 @@
+  }
+  std::shared_ptr <std::unordered_map<int, TVector3>> MapPointer;
+  MapPointer = std::make_shared<std::unordered_map<int, TVector3>>(positionMap);
+  return MapPointer;
```
wouldn't `return std::make_shared<std::unordered_map<int, TVector3>>(positionMap);` be sufficient?

owtscharenko replied:
I guess you are right, this might be a left over from former implementations

olantwin commented:
Looks good to me code-wise.

Can't test right now though. Does it work as expected?

owtscharenko commented:
I would not know how to test...

olantwin commented:
For simple testing, just checking whether the map is created when expected and whether the position can be retrieved for some hits should be fine.

owtscharenko commented:
yes, but I have no Idea how to do that. I guess we would need a reconstruction script?

olantwin commented:
Why? The hits we can use from the raw data that we convert.

But yes, we should in general modify the muon flux reconstruction script to suit the needs of the charm measurement.
